contracts/token/ERC20/extensions/ERC1363.sol

```diff
@@ -0,0 +1,142 @@
+    function transferAndCall(
+        address to,
+        uint256 value,
+        bytes memory data
+    ) public override returns (bool) {
+        require(transfer(to, value));
+        require(
+            _checkOnTransferReceived(_msgSender(), to, value, data),
+            "ERC1363: transfer to non ERC1363Receiver implementer"
+        );
+        return true;
+    }
+
+    /**
+     * @dev See {IERC1363-transferFromAndCall}.
+     */
+    function transferFromAndCall(
+        address from,
+        address to,
+        uint256 value
+    ) public override returns (bool) {
+        return transferFromAndCall(from, to, value, bytes(""));
+    }
+
+    /**
+     * @dev See {IERC1363-transferFromAndCall}.
+     */
+    function transferFromAndCall(
+        address from,
+        address to,
+        uint256 value,
+        bytes memory data
+    ) public override returns (bool) {
+        require(transferFrom(from, to, value));
+        require(
+            _checkOnTransferReceived(from, to, value, data),
+            "ERC1363: transfer to non ERC1363Receiver implementer"
+        );
+        return true;
+    }
+
+    /**
+     * @dev See {IERC1363-approveAndCall}.
+     */
+    function approveAndCall(address spender, uint256 value) public override returns (bool) {
+        return approveAndCall(spender, value, bytes(""));
+    }
+
+    /**
+     * @dev See {IERC1363-approveAndCall}.
+     */
+    function approveAndCall(
+        address spender,
+        uint256 value,
+        bytes memory data
+    ) public override returns (bool) {
+        require(approve(spender, value));
+        require(
+            _checkOnApprovalReceived(_msgSender(), spender, value, data),
+            "ERC1363: transfer to non ERC1363Spender implementer"
+        );
+        return true;
+    }
+
+    /**
+     * @dev Internal function to invoke {IERC1363Receiver-onTransferReceived} on a target address.
+     * The call is not executed if the target address is not a contract.
+     */
+    function _checkOnTransferReceived(
+        address from,
+        address to,
+        uint256 value,
+        bytes memory data
+    ) private returns (bool) {
+        try IERC1363Receiver(to).onTransferReceived(_msgSender(), from, value, data) returns (bytes4 retval) {
+            return retval == IERC1363Receiver.onTransferReceived.selector;
+        } catch (bytes memory reason) {
+            if (reason.length == 0) {
+                revert("ERC1363: transfer to non ERC1363Receiver implementer");
+            } else {
+                /// @solidity memory-safe-assembly
+                assembly {
+                    revert(add(32, reason), mload(reason))
+                }
+            }
+        }
+    }
+
+    /**
+     * @dev Internal function to invoke {IERC1363Spender-onApprovalReceived} on a target address.
+     * The call is not executed if the target address is not a contract.
+     */
+    function _checkOnApprovalReceived(
+        address owner,
```

Comment:
Why passing owner here and not using msg.sender?

Reply:
It is in our best practices that the internal and private function should be context-free, and that the public/external interface should extract the context and pass it through arguments.

Reply:
Ok, so should `_msgSender()` also to be removed from there?
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/eadcad5e6945462b3e1d056ec770353207ed57c5/contracts/token/ERC20/extensions/ERC1363.sol#L105

Reply:
yes